### PR TITLE
Add derivatives for mvn

### PR DESCRIPTION
**Pull request by spinkney**
## Summary

This PR adds derivatives for the `multi_normal_lpdf`. 

## Tests

Added the derivative tests in `test/unit/math/rev/prob/multi_normal_test.cpp`

## Side Effects

No

## Release notes

Added hand-calculated derivatives for the multivariate normal lpdf.

## Checklist

- [x] Copyright holder: Sean Pinkney

    The copyright holder is typically you or your assignee, such as a university or company. By submitting this pull request, the copyright holder is agreeing to the license the submitted work under the following licenses:
      - Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
      - Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

- [x] the basic tests are passing

    - unit tests pass (to run, use: `./runTests.py test/unit`)
    - header checks pass, (`make test-headers`)
    - dependencies checks pass, (`make test-math-dependencies`)
    - docs build, (`make doxygen`)
    - code passes the built in [C++ standards](https://github.com/stan-dev/stan/wiki/Coding-Style-and-Idioms) checks (`make cpplint`)

- [x] the code is written in idiomatic C++ and changes are documented in the doxygen

- [x] the new changes are tested

### CODE REVIEW

**Comment by spinkney** [comment]
The derivative tests fail with 
```cpp
./stan/math/prim/prob/multi_normal_lpdf.hpp:237:11: note: in instantiation of function template specialization 'Eigen::Matrix<double, -1, 1>::operator=<Eigen::Matrix<stan::math::var_value<double>, -1, 1>>' requested here
     half = mdivide_left_ldlt(ldlt_Sigma, y_val_minus_mu_val);
```
The offending code is at https://github.com/stan-dev/math/blob/208658abe91cc8dc36c77ca240a52f3ad7fcfc25/stan/math/prim/prob/multi_normal_lpdf.hpp#L239. I believe the `ldlt_factor` needs to be a `double` but it's passing a `const stan::math::var_value<double>`. This is beyond my current cpp skills.

File: stan/math/prim/prob/multi_normal_lpdf.hpp [hunk 1]
```diff
@@ -71,35 +90,173 @@ return_type_t<T_y, T_loc, T_covar> multi_normal_lpdf(const T_y& y,
...
+  check_size_match(function, "Size of random variable", size_y,
+                   "size of location parameter", size_mu);
+  check_size_match(function, "Size of random variable", size_y,
+                   "rows of covariance parameter", Sigma.rows());
+  check_size_match(function, "Size of random variable", size_y,
+                   "columns of covariance parameter", Sigma.cols());
+
+  check_symmetric(function, "Covariance matrix", Sigma_ref);
+
+  auto ldlt_Sigma = make_ldlt_factor(Sigma_ref);
```
Comment (WardBrian) [inline stan/math/prim/prob/multi_normal_lpdf.hpp:207]: I think this line is where you're getting vars that lead to an error later. At least for the first derivative, the `partial_return_t` type is going to be `double`, but `Sigma_ref` might be a `var`. 

`auto ldlt_Sigma = make_ldlt_factor(value_of(Sigma_ref));` gets your test compiling (and may even be correct?).

Comment (WardBrian) [reply stan/math/prim/prob/multi_normal_lpdf.hpp]: It looks like every single place we call `make_ldlt_factor` we use auto, so I doubt we're the only ones who don't know how to name it's type generically

**Comment by spinkney** [comment]
Benchmarks

Slower for size 1 and 2, about the same for 4, and starts to get faster at 8 and up.

```
Base (Develop branch): Failed after size 1024
--------------------------------------------------------------------------------------------------------------
Benchmark                                                                    Time             CPU   Iterations
--------------------------------------------------------------------------------------------------------------
multi_normal_lpdf_Mix_vector_Mix_vector_Mix_matrix/1/manual_time           501 ns          657 ns      1198449
multi_normal_lpdf_Mix_vector_Mix_vector_Mix_matrix/2/manual_time           828 ns         1014 ns       849637
multi_normal_lpdf_Mix_vector_Mix_vector_Mix_matrix/4/manual_time          1951 ns         2267 ns       349908
multi_normal_lpdf_Mix_vector_Mix_vector_Mix_matrix/8/manual_time          6789 ns         7611 ns       104321
multi_normal_lpdf_Mix_vector_Mix_vector_Mix_matrix/16/manual_time        34050 ns        36982 ns        20209
multi_normal_lpdf_Mix_vector_Mix_vector_Mix_matrix/32/manual_time       190091 ns       201052 ns         3802
multi_normal_lpdf_Mix_vector_Mix_vector_Mix_matrix/64/manual_time      1162380 ns      1204392 ns          590
multi_normal_lpdf_Mix_vector_Mix_vector_Mix_matrix/128/manual_time     8196214 ns      8355549 ns           82
multi_normal_lpdf_Mix_vector_Mix_vector_Mix_matrix/256/manual_time    66225616 ns     66894889 ns            9
multi_normal_lpdf_Mix_vector_Mix_vector_Mix_matrix/512/manual_time  1944669500 ns   1808821000 ns            1
multi_normal_lpdf_Mix_vector_Mix_vector_Mix_matrix/1024/manual_time 23493403750 ns   19443533000 ns            1


New (multi_normal_lpdf with derivatives)
--------------------------------------------------------------------------------------------------------------
Benchmark                                                                    Time             CPU   Iterations
--------------------------------------------------------------------------------------------------------------
multi_normal_lpdf_Mix_vector_Mix_vector_Mix_matrix/1/manual_time          1197 ns         1444 ns       534847
multi_normal_lpdf_Mix_vector_Mix_vector_Mix_matrix/2/manual_time          1445 ns         1722 ns       488717
multi_normal_lpdf_Mix_vector_Mix_vector_Mix_matrix/4/manual_time          2052 ns         2469 ns       344037
multi_normal_lpdf_Mix_vector_Mix_vector_Mix_matrix/8/manual_time          4212 ns         5126 ns       169123
multi_normal_lpdf_Mix_vector_Mix_vector_Mix_matrix/16/manual_time        13062 ns        15986 ns        51234
multi_normal_lpdf_Mix_vector_Mix_vector_Mix_matrix/32/manual_time        49680 ns        60607 ns        14207
multi_normal_lpdf_Mix_vector_Mix_vector_Mix_matrix/64/manual_time       200406 ns       243625 ns         3450
multi_normal_lpdf_Mix_vector_Mix_vector_Mix_matrix/128/manual_time      884341 ns      1057812 ns          767
multi_normal_lpdf_Mix_vector_Mix_vector_Mix_matrix/256/manual_time     4661935 ns      5350265 ns          151
multi_normal_lpdf_Mix_vector_Mix_vector_Mix_matrix/512/manual_time    24922471 ns     27647185 ns           27
multi_normal_lpdf_Mix_vector_Mix_vector_Mix_matrix/1024/manual_time  157399000 ns    168161750 ns            4
multi_normal_lpdf_Mix_vector_Mix_vector_Mix_matrix/2048/manual_time 1741480334 ns   1697054000 ns            1
multi_normal_lpdf_Mix_vector_Mix_vector_Mix_matrix/4096/manual_time 12310096541 ns   11754263000 ns         
```

**Comment by spinkney** [comment]
Somewhat surprising to me is that this ends up benchmarking faster than `multi_normal_cholesky_lpdf` for the mix tests.

I.e.
`python3 ./benchmarks/benchmark.py "multi_normal_cholesky_lpdf(real, real, matrix) => real" --skip_similar_signatures --overloads Mix`

**Comment by spinkney** [comment]
@bgoodri this doesn't benchmark any faster than the current autodiff code. Does it make sense to add this for stability or should we just keep the current autodiffed code?

**Comment by bgoodri** [comment]
> @bgoodri this doesn't benchmark any faster than the current autodiff code. Does it make sense to add this for stability or should we just keep the current autodiffed code?

I would say that if it is no slower, then having the analytical derivatives is worth it because it consumes less RAM.

**Comment by spinkney** [comment]
@SteveBronder this should be good for a review once the CI finishes

**Review by SteveBronder — DISMISSED** [review]

**Comment by syclik** [comment]
@spinkney, just letting you know there are merge conflicts; @SteveBronder approved the PR!

**Comment by spinkney** [comment]
@serban-nicusor-toptal does CI need to be restarted? I've never seen this error before.

**Comment by serban-nicusor-toptal** [comment]
Jenkins shenanigans, everything looks fine now!

**Review by SteveBronder — APPROVED** [review]

**Pull request merged**